swarm/api/api.go

```diff
@@ -72,21 +72,46 @@ type ErrResolve error
 // DNS Resolver
 func (self *Api) Resolve(hostPort string, nameresolver bool) (storage.Key, error) {
 	log.Trace(fmt.Sprintf("Resolving : %v", hostPort))
-	if hashMatcher.MatchString(hostPort) || self.dns == nil {
-		log.Trace(fmt.Sprintf("host is a contentHash: '%v'", hostPort))
-		return storage.Key(common.Hex2Bytes(hostPort)), nil
-	}
-	if !nameresolver {
-		return nil, fmt.Errorf("'%s' is not a content hash value.", hostPort)
+
+	//if self.dns is nil, don't resolve, but first try fall back to hash
+	if self.dns == nil {
```

Comment:
I think this is all you need for the resolve/fallback logic, then keep the functions as is and keep raw as nameresolve=true. All the tests that call with a proper hash now go through. 
We need to add tests to test these error responses and then we are done :)s